core/evm.go

```diff
@@ -501,3 +527,183 @@ func Transfer(db vm.StateDB, sender, recipient common.Address, amount *big.Int,
+						delegation.DelegatorAddress[:],
+						roStakeMsg.DelegatorAddress[:],
+					) {
+						return common.LeftPadBytes(
+							delegation.Amount.Bytes(), 32,
+						), nil
+					}
+				}
+				return big.NewInt(0).Bytes(), nil
+			}
+		case "ValidatorMaxTotalDelegation":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				return common.LeftPadBytes(
+					wrapper.MaxTotalDelegation.Bytes(), 32,
+				), nil
+			}
+		case "ValidatorTotalDelegation":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				return common.LeftPadBytes(
+					wrapper.TotalDelegation().Bytes(), 32,
+				), nil
+			}
+		case "BalanceAvailableForRedelegation":
+			{
+				delegations, err := chain.ReadDelegationsByDelegatorAt(
+					roStakeMsg.DelegatorAddress,
+					big.NewInt(0).Sub(ref.Number(), big.NewInt(1)),
+				)
+				if err != nil {
+					return nil, err
+				}
+				redelegationTotal := big.NewInt(0)
+				for _, delegationIndex := range delegations {
+					wrapper, err := db.ValidatorWrapper(delegationIndex.ValidatorAddress, true, false)
+					if err != nil {
+						return nil, err
+					}
+					if uint64(len(wrapper.Delegations)) > delegationIndex.Index {
+						delegation := wrapper.Delegations[delegationIndex.Index]
+						for _, undelegation := range delegation.Undelegations {
+							if undelegation.Epoch.Cmp(ref.Epoch()) < 1 { // Undelegation.Epoch < currentEpoch
+								redelegationTotal.Add(redelegationTotal, undelegation.Amount)
+							}
+						}
+					}
+				}
+				return common.LeftPadBytes(
+					redelegationTotal.Bytes(), 32,
+				), nil
+			}
+		case "ValidatorCommissionRate":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				// Solidity doesn't support float directly, so just return the bytes
+				// the bytes can't be an exact number but can be used for comparison
+				// for example, a 1.1 commission rate shows up as 1100000000000000000
+				// divide by 1e18 (in Python) to get the actual rate 1.1
+				return common.LeftPadBytes(
+					wrapper.CommissionRates.Rate.Bytes(), 32,
+				), nil
+			}
+		case "ValidatorStatus":
+			{
+				wrapper, err := db.ValidatorWrapper(roStakeMsg.ValidatorAddress, true, false)
+				if err != nil {
+					return nil, err
+				}
+				return wrapper.Status.Bytes(), nil
+			}
+		case "BalanceDelegatedByDelegator":
+			{
+				delegations, err := chain.ReadDelegationsByDelegator(roStakeMsg.DelegatorAddress)
+				if err != nil {
+					return nil, err
+				}
+				answer := big.NewInt(0)
+				for _, delegation := range delegations {
+					wrapper, err := db.ValidatorWrapper(delegation.ValidatorAddress, true, false)
+					if err != nil {
+						return nil, err
+					}
+					answer = answer.Add(answer, wrapper.Delegations[delegation.Index].Amount)
```

Comment:
Please add a index oob check. Even though it shouldn't happen, it's safe to do that.